File: client/src/main/java/com/flipkart/grayskull/utils/RetryUtil.java
package com.flipkart.grayskull.utils;

import com.flipkart.grayskull.models.exceptions.GrayskullException;
import com.flipkart.grayskull.models.exceptions.RetryableException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.UUID;

/**
 * Utility class for retrying operations with exponential backoff.
 * <p>
 * This class implements a retry mechanism with exponential backoff for operations
 * that may fail transiently. It will retry operations that throw {@link RetryableException}
 * up to a maximum number of attempts, with increasing delays between attempts.
 * The wait time is capped at a maximum of 1 minute.
 * </p>
 */
public final class RetryUtil {
    private static final Logger log = LoggerFactory.getLogger(RetryUtil.class);
    private static final long MAX_WAIT_TIME_MS = 60000; // 1 minute

    private final int maxAttempt;
    private final int interval;

    /**
     * Constructor to initialize the RetryUtil with the number of attempts and the interval between retries.
     *
     * @param maxAttempt The maximum number of retry attempts.
     * @param interval   The initial interval between retries in milliseconds.
     */
    public RetryUtil(int maxAttempt, int interval) {
        this.maxAttempt = maxAttempt;
        this.interval = interval;
    }

    /**
     * Executes the given task with retry logic.
     * <p>
     * If the task throws a {@link RetryableException}, it will be retried up to maxAttempt times
     * with exponential backoff. If all retry attempts are exhausted, the {@link RetryableException}
     * is wrapped in a {@link GrayskullException}. Non-retryable exceptions are thrown immediately.
     * The wait time between retries is capped at a maximum of 1 minute.
     * </p>
     *
     * @param task The task to execute
     * @param <T>  The return type of the task
     * @return The result of the task
     * @throws GrayskullException if the task fails after all retry attempts
     * @throws Exception if the task throws a non-retryable exception
     */
    public <T> T retry(CheckedSupplier<T> task) throws Exception {
        // Generate a short retry ID to track this retry sequence in logs
        String retryId = UUID.randomUUID().toString().substring(0, 8);
        long currentInterval = interval;
        Exception lastException = null;
        
        for (int attempt = 1; attempt <= maxAttempt; attempt++) {
            try {
                log.debug("[retry:{}] Executing task, attempt {} of {}", retryId, attempt, maxAttempt);
                T result = task.get();
                if (attempt > 1) {
                    log.info("[retry:{}] Task succeeded on attempt {}", retryId, attempt);
                }
                return result;
            } catch (RetryableException e) {
                lastException = e;
                log.warn("[retry:{}] Retryable exception on attempt {} of {}: {}", retryId, attempt, maxAttempt, e.getMessage());
                
                if (attempt == maxAttempt) {
                    log.error("[retry:{}] Max retry attempts reached ({}), throwing exception", retryId, maxAttempt);
                    throw new GrayskullException(e.getStatusCode(), "Failed after " + maxAttempt + " retry attempts: " + e.getMessage(), e);
                }
                
                // Sleep before next retry (exponential backoff)
                try {
                    log.debug("[retry:{}] Waiting {}ms before retry", retryId, currentInterval);
                    Thread.sleep(currentInterval);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    throw new GrayskullException("Retry interrupted", ie);
                }
                currentInterval = Math.min(currentInterval * 2, MAX_WAIT_TIME_MS); // Exponential backoff capped at 1 minute
            } 
        }
        throw new GrayskullException("Reached maximum attempts: " + maxAttempt, lastException);
    }

    /**
     * Functional interface representing a supplier that can throw an exception.
     *
     * @param <T> The type of the result supplied by this supplier.
     */
    @FunctionalInterface
    public interface CheckedSupplier<T> {
        T get() throws Exception;
    }
}

File: clients/go/client-impl/utils/retryUtil.go
package utils

import (
	"context"
	"errors"
	"math"
	"math/rand"
	"time"

	"github.com/flipkart-incubator/grayskull/client-impl/models/exceptions"
)

// RetryConfig holds configuration for retry behavior
type RetryConfig struct {
	MaxAttempts   int
	InitialDelay  time.Duration
	MaxRetryDelay time.Duration
}

// RetryUtil provides retry functionality with exponential backoff
type RetryUtil struct {
	config RetryConfig
}

// NewRetryUtil creates a new RetryUtil with the given configuration
func NewRetryUtil(config RetryConfig) *RetryUtil {
	if config.MaxAttempts <= 0 {
		config.MaxAttempts = 3
	}
	if config.InitialDelay <= 0 {
		config.InitialDelay = 100 * time.Millisecond
	}
	if config.MaxRetryDelay <= 0 {
		config.MaxRetryDelay = time.Minute
	}

	return &RetryUtil{
		config: config,
	}
}

// Retry executes the task with retry logic
func (r *RetryUtil) Retry(ctx context.Context, task func() (interface{}, error)) (interface{}, error) {
	var lastErr error
	delay := r.config.InitialDelay

	for attempt := 1; attempt <= r.config.MaxAttempts; attempt++ {
		// Check if context is done before each attempt
		if err := ctx.Err(); err != nil {
			return nil, exceptions.NewRetryableErrorWithCause("operation canceled", err)
		}

		// Execute the task
		result, err := task()
		if err == nil {
			return result, nil
		}

		// Check if error is retryable
		var retryableErr *exceptions.RetryableError
		if !errors.As(err, &retryableErr) {
			return nil, err // Non-retryable error
		}

		lastErr = retryableErr
		if attempt >= r.config.MaxAttempts {
			break
		}

		// Calculate backoff with jitter
		delay = time.Duration(
			math.Min(
				float64(r.config.InitialDelay)*math.Pow(2, float64(attempt-1)),
				float64(r.config.MaxRetryDelay),
			),
		)
		// Add jitter (up to 25% of the delay)
		jitter := time.Duration(rand.Float64() * 0.25 * float64(delay))
		delay += jitter

		// Wait before next retry
		select {
		case <-time.After(delay):
		case <-ctx.Done():
			return nil, exceptions.NewRetryableErrorWithCause("operation canceled during retry", ctx.Err())
		}
	}

	// If we've exhausted all retry attempts, return a GrayskullError
	return nil, exceptions.NewGrayskullErrorWithMessageAndCause(
		"max retry attempts reached",
		lastErr,
	)
}
